### PR TITLE
Add model downsampling script

### PR DESCRIPTION
**Pull request by younesbelkada**
What does this PR do?

Adds model downsampling scripts to be able to try model shrinking experiments!
Note that it is a WIP draft PR for now

### CODE REVIEW

File: downsample_model.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    return sum(p.numel() for p in model.parameters() if p.requires_grad)
+
+def convert_config_to_downsized_config(config, downsampling_rate):
+    """
+        Step1: Define the new hyperparameters given the source model and the downsampling rate. This can be done in the config file
+    """
+    old_hidden_size, old_num_layer = config.hidden_size, config.n_layer
+    downsized_config = deepcopy(config)
+    downsized_config.hidden_size = int(old_hidden_size * downsampling_rate * 2)
+    downsized_config.n_layer = int(old_num_layer * downsampling_rate * 2)
```
Comment (ibeltagy) [inline downsample_model.py]: you should have a different rate for depth and with

Comment (ibeltagy) [reply downsample_model.py]: when you change hidden size, do you also need to change number of heads?

File: downsample_model.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+        splitting_dimensions = (4*hidden_size,)
+    elif key.endswith("dense_4h_to_h.weight"):
+        splitting_dimensions = (hidden_size, 4*hidden_size)
+    elif key.endswith("self_attention.query_key_value.bias"):
+        splitting_dimensions = (3*hidden_size,)
+    elif key.endswith("self_attention.query_key_value.weight"):
+        splitting_dimensions = (3*hidden_size, hidden_size)
+    else:
+        splitting_dimensions = (hidden_size,)
+    return splitting_dimensions
```
Comment (ibeltagy) [inline downsample_model.py]: I am not following how this function works. Doesn't this function need the downsampling rate as an arg?

File: downsample_model.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+        return splitted_tensor[0]
+    elif aggregation_strategy == "last":
+        return splitted_tensor[-1]
+    else:
+        raise ValueError("Unknown aggregation strategy: {}".format(aggregation_strategy))
+
+def downsize_state_dict(state_dict, downsized_model_config, downsampling_rate, aggregation_strategy="mean"):
+    """
+        The state dict needs to be pre-processed by popping the keys that are not needed (aka the layers that we want to discard)
+        Step5: Load the source model's weights into the new model and save the new model. Let the user define the aggregation strategy
```
Comment (ibeltagy) [inline downsample_model.py]: the code so far only works for width, right? I don't seem to be able to find the code for downsampling layers

File: downsample_model.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    return splitting_dimensions
+
+
+def process_weights(weights, aggregation_strategy, downsampling_rate, splitting_dimensions=1):
+    """
+        This function takes the weights of the old model and downsamples them to the new model's dimensions
+    """
+    if len(splitting_dimensions) == 1:
+        splitting_size = splitting_dimensions[0]
+        splitted_tensor = torch.stack(weights.split(splitting_size, dim=-1))
```
Comment (ibeltagy) [inline downsample_model.py]: same - can you explain why this is not using the downsampling rate?

**Review by ibeltagy — COMMENTED** [review]
left a few clarification questions. Thanks, @younesbelkada

File: downsample_model.py (hunk 3 above)
Comment (younesbelkada) [reply downsample_model.py]: yeah I just forgot to push it ! Sorry for that

Comment (younesbelkada) [reply downsample_model.py]: should be fixed in 5ed4b61cd69546e2235e87dedc50a4afafaf5fa9

File: downsample_model.py (hunk 4 above)
Comment (younesbelkada) [reply downsample_model.py]: the variable `splitting_size` corresponds to the desired shape of the target weight matrices. This variable implicitly uses `downsampling_rate`. 
I have updated the docstring accordingly and factored a bit the code to make it more clear

File: downsample_model.py (hunk 1 above)
Comment (younesbelkada) [reply downsample_model.py]: Yes agreed - should be fixed in 5738abbcd043868abd6b3968aea8fe21bfddeeb7

Comment (younesbelkada) [reply downsample_model.py]: From slack I thought that we keep the number of heads untouched

File: downsample_model.py (hunk 2 above)
Comment (younesbelkada) [reply downsample_model.py]: Same as below, the downsampling rate is implicitly used when creating `hidden_size`, I should have updated the docstring to make it more clear, see 5738abbcd043868abd6b3968aea8fe21bfddeeb7

**Comment by younesbelkada** [comment]
Hi all! 
I marked this PR as ready for review! Should be usable and follows what I have understood from our discussions 
@ibeltagy @justheuristic 
Let me know if you see anything that needs to be changed there!

File: downsample_model.py [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+def main(args):
+    model_name = args.model_name
+    output_model_name = args.output_model_name
+    hidden_downsampling_rate = args.hidden_downsampling_rate
+    layer_downsampling_rate = args.layer_downsampling_rate
+    aggregation_strategy = args.aggregation_strategy
+    layer_selection_strategy = args.layer_selection_strategy
+    push_to_hub = args.push_to_hub
+    
+    config_old_model = BloomConfig.from_pretrained(model_name, use_auth_token=True)
```
Comment (ibeltagy) [inline downsample_model.py]: nit: just use `args.model_name`. Same for other args. Less code and less error-prone.

File: downsample_model.py [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+def convert_config_to_downsized_config(config, hidden_size_downsampling_rate, width_downsampling_rate, aggregation_strategy, layer_selection_strategy):
+    """
+        Defines the new hyperparameters of the shrinked model, given the downsampling rates.
+        Creates a new config file and returns it.
+        Args:
+            config (`transformers.BloomConfig`, *required*):
+                input old model's config
+            hidden_downsampling_rate (`float`, *required*):
+                hidden states downsampling rate
+            width_downsampling_rate (`float`, *required*):
```
Comment (ibeltagy) [inline downsample_model.py]: I think you mean `depth_downsampling_rate`

**Review by ibeltagy — COMMENTED** [review]
Looks great. Thanks, @younesbelkada. I left a couple of nits. 

Also, can you please add unit tests? The functionality implemented here is nuanced enough that unit tests can save us a lot of headaches down the line.

File: downsample_model.py (hunk 6 above)
Comment (younesbelkada) [reply downsample_model.py]: yeah true! thanks for the nit
FIxed in 854dccf92fee29818a6cfc69e6ebc6e549a6ac7f

File: downsample_model.py (hunk 5 above)
Comment (younesbelkada) [reply downsample_model.py]: Perfect! Fixed in 854dccf92fee29818a6cfc69e6ebc6e549a6ac7f

**Comment by younesbelkada** [comment]
I propose to refactor the code a bit and add relevant tests in e09a54a9e0ac94184dd7e45953f2d3a0bb40e70f
Let me know what do you think !

File: src/downsampling.py [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+        {source_model_key:target_model_key, ... }
+        Note that this is applied only on the attention block layers (named as `h` in `BloomModel`)
+
+        Example:
+            downsampling rate = 0.5 (ie consider half of the layers)
+            Source Model: 
+                h.0 | h.1 | ... | h.30
+            Target Model:
+                if strategy = 'first':
+                    returns {h.0:h.0, h.1:h.1, ... h.14:h.14} (consider half of the layers)
```
Comment (younesbelkada) [inline src/downsampling.py]:
```suggestion
                    returns {h.0:h.0, h.1:h.1, ... h.14:h.14} (consider the first half of the layers)
```

File: src/downsampling.py [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+                if strategy = 'last':
+                    returns {h.15:h.0, h.16:h.0, ... , h.30:h.14}
```
Comment (younesbelkada) [inline src/downsampling.py]:
```suggestion
                    returns {h.15:h.0, h.16:h.0, ... , h.30:h.14} (consider the second half of the layers
```

File: src/downsampling.py [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+
+        Example:
+            downsampling rate = 0.5 (ie consider half of the layers)
+            Source Model: 
+                h.0 | h.1 | ... | h.30
+            Target Model:
+                if strategy = 'first':
+                    returns {h.0:h.0, h.1:h.1, ... h.14:h.14} (consider the first half of the layers)
+                if strategy = 'last':
+                    returns {h.15:h.0, h.16:h.0, ... , h.30:h.14} (consider the second half of the layers
```
Comment (younesbelkada) [inline src/downsampling.py]:
```suggestion
                    returns {h.15:h.0, h.16:h.0, ... , h.30:h.14} (consider the second half of the layers)
```

**Pull request merged**